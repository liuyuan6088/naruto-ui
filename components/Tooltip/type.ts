
import { PlacementType, Trigger } from '../Portal/type'

export interface ITooltipProps {
  prefixCls?: string;
  // style dropDown 样式
  dropDownStyle?: React.CSSProperties
  // dropDown className
  dropDownClassName?: string
  // 提示文字
  title: React.ReactNode;
  // 弹出方向
  placement?: PlacementType;
  // offset 偏移距离 默认为6
  offset?: number
  // onVisibleChange 显隐状态变化的回调
  onVisibleChange?: (visible: boolean) => void
  // visible 是否可见 受控
  visible?: boolean;
  // 触发 dropdown 的方式
  trigger?: Trigger;
  // 鼠标移入后延时多少才显示 Tooltip 单位：秒
  mouseEnterDelay?: number;
  // 鼠标移出后延时多少才隐藏 Tooltip 单位：秒
  mouseLeaveDelay?: number;
  // 动画的类名
  transitionName?: string;
}