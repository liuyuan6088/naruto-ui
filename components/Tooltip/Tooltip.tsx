import * as React from 'react'
import Portal from '../Portal'
import { ITooltipProps } from './type'
import { primaryName } from '../utils/constant'
import cx from 'classnames'
import './style/index.less'

const defaultProps: ITooltipProps = {
  prefixCls: `${primaryName}-tooltip`,
  placement: 'top',
  offset: 0,
  trigger: 'hover',
  title: '',
  mouseEnterDelay: 0.1,
  mouseLeaveDelay: 0.1
}

const Tooltip: React.FC<ITooltipProps> = props => {
  const {
    prefixCls,
    visible,
    trigger,
    placement,
    offset,
    onVisibleChange,
    title,
    children,
    ...rest
  } = props

  const classes = cx({
    [`${prefixCls}-${placement}`]: placement
  })

  const content = <div className={`${prefixCls}-content-inner`}>{title}</div>

  return (
    <Portal
      dropDownClassName={classes}
      prefixCls={prefixCls}
      visible={visible}
      content={content}
      transitionName="zoom-big-fast"
      offset={offset}
      placement={placement}
      trigger={trigger}
      onVisibleChange={onVisibleChange}
      wrapperStyle={{ display: 'inline-block' }}
      hasTriangle
      {...rest}
    >
      {children}
    </Portal>
  )
}

Tooltip.defaultProps = defaultProps

export default Tooltip
