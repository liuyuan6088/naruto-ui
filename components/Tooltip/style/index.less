@import '../../styles/theme/default.less';
// @import '../../styles/mixins/index.less';

@tooltip-prefix-cls: ~"@{primary-name}-tooltip";

.@{tooltip-prefix-cls} {
  position: absolute;
  z-index: @tooltip-basic-zIndex;
  font-size: 14px;
  font-variant: tabular-nums;
  line-height: 1.5;
  list-style: none;
  cursor: pointer;

  &-wrapper {
    display: inline-block;
    cursor: pointer;
  }

  &-content {
    color: #fff;
    background-color: @tooltip-content-bgc;
    border-radius: @bordor-radius;
    box-shadow: @tooltip-content-shadow;
    min-width: 30px;
    min-height: 32px;
    padding: 6px 8px;
    color: #fff;
    text-align: left;
    text-decoration: none;
    word-wrap: break-word;
    position: relative;
    cursor: pointer;
  }

  &-triangle {
    .@{tooltip-prefix-cls}-content {
      &::after {
        content: '';
        position: absolute;
        width: 0;
        height: 0;
        border-color: transparent;
        border-style: solid;
        box-sizing: border-box;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-top,
  &-triangle.@{tooltip-prefix-cls}-topLeft,
  &-triangle.@{tooltip-prefix-cls}-topRight {
    padding-bottom: 8px;
    .@{tooltip-prefix-cls}-content {
      &::after {
        bottom: -5px;
        border-width: 5px 5px 0;
        border-top-color: @tooltip-content-bgc;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-bottom,
  &-triangle.@{tooltip-prefix-cls}-bottomLeft,
  &-triangle.@{tooltip-prefix-cls}-bottomRight {
    padding-top: 8px;
    .@{tooltip-prefix-cls}-content {
      &::after {
        top: -5px;
        border-width: 0 5px 5px;
        border-bottom-color: @tooltip-content-bgc;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-left,
  &-triangle.@{tooltip-prefix-cls}-leftTop,
  &-triangle.@{tooltip-prefix-cls}-leftBottom {
    padding-right: 8px;
    .@{tooltip-prefix-cls}-content {
      &::after {
        right: -5px;
        border-width: 5px 0 5px 5px;
        border-left-color: @tooltip-content-bgc;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-right,
  &-triangle.@{tooltip-prefix-cls}-rightTop,
  &-triangle.@{tooltip-prefix-cls}-rightBottom {
    padding-left: 8px;
    .@{tooltip-prefix-cls}-content {
      &::after {
        left: -5px;
        border-width: 5px 5px 5px 0;
        border-right-color: @tooltip-content-bgc;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-top,
  &-triangle.@{tooltip-prefix-cls}-bottom {
    .@{tooltip-prefix-cls}-content {
      &::after {
        left: 50%;
        transform: translateX(-50%);
      }
    }
  }
  &-triangle.@{tooltip-prefix-cls}-topLeft,
  &-triangle.@{tooltip-prefix-cls}-bottomLeft {
    .@{tooltip-prefix-cls}-content {
      &::after {
        left: 16px;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-topRight,
  &-triangle.@{tooltip-prefix-cls}-bottomRight {
    .@{tooltip-prefix-cls}-content {
      &::after {
        right: 16px;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-leftTop,
  &-triangle.@{tooltip-prefix-cls}-rightTop {
    .@{tooltip-prefix-cls}-content {
      &::after {
        top: 8px;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-left,
  &-triangle.@{tooltip-prefix-cls}-right {
    .@{tooltip-prefix-cls}-content {
      &::after {
        top: 50%;
        margin-top: -5px;
      }
    }
  }

  &-triangle.@{tooltip-prefix-cls}-leftBottom,
  &-triangle.@{tooltip-prefix-cls}-rightBottom {
    .@{tooltip-prefix-cls}-content {
      &::after {
        bottom: 8px;
      }
    }
  }
}