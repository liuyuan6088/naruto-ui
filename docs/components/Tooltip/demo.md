---
imports:
  import Demo from './demo.tsx';
  import DemoCode from '!raw-loader!./demo.tsx';

---

## Spin 加载中

用于页面和区块的加载中状态。

## 使用加载中

:::demo
  <CodeBox code={DemoCode}>
    <Demo />
  </CodeBox>
:::

## 静态方法

```js
// Spin.setDefaultIndicator(indicator: ReactElement)
// 同上 indicator，你可以自定义全局默认元素
Spin.setDefaultIndicator(
  <Icon type='loading' style={{ color: 'red' }} spin />
)
```

## Props
| 参数     | 说明     | 类型   | 可选值 | 默认值 |
| -------- | -------- | ------ | ------ | ------ |
| tip | 当作为包裹元素时，可以自定义描述文案 | string / React.ReactNode | - | - |
| size | 组件大小 | string / `large` /  `default` / `small` | - |
| spinning | 是否为加载中状态 | boolean | - | true |
| indicator | 加载指示符 | React.ReactNode | - | - |
| delay | 延迟显示加载效果的时间（防止闪烁） | number | - | - |
| className | 类名 | string | - | - |
| style | 样式 | React.CSSProperties | - | - |

