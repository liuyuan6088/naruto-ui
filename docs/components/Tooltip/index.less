.demo-tooltip-1 {
  position: relative;
  height: 250px;
  width: 700px;
  margin: 0 auto;
  button {
    width: 110px;
  }
  .demo-tooltip-top {
    position: absolute;
    top: 0;
    left: 50%;
    transform: translateX(-50%);
    z-index: 10;
    > div {
      & + div {
        margin-left: 10px;
      }
    }
  }
  .demo-tooltip-left {
    position: absolute;
    width: 110px;
    left: 0;
    top: 50%;
    transform: translateY(-50%);
    > div {
      & + div {
        margin-top: 10px;
      }
    }
  }
  .demo-tooltip-right {
    position: absolute;
    right: 0;
    width: 110px;
    top: 50%;
    transform: translateY(-50%);
    > div {
      & + div {
        margin-top: 10px;
      }
    }
  }
  .demo-tooltip-bottom {
    position: absolute;
    bottom: 0;
    left: 50%;
    transform: translateX(-50%);
    > div {
      & + div {
        margin-left: 10px;
      }
    }
  }
}