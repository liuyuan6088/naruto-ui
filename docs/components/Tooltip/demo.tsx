import * as React from 'react';
import { Tooltip, Button } from 'components';
import './index.less'

const topList = ['topLeft', 'top', 'topRight']
const leftList = ['leftTop', 'left', 'leftBottom']
const bottomList = ['bottomLeft', 'bottom', 'bottomRight']
const rightList = ['rightTop', 'right', 'rightBottom']

const Demo: React.FC = () => {

  const [visible, setVisible] = React.useState<boolean>(false)

  return (
    <React.Fragment>

      <p>基本用法</p>
      <div className='demo-spin-1'>
        <Tooltip title='Tooltip!!!!!!!!!' onVisibleChange={bool => console.log(bool)}>
          <Button>基本用法</Button>
        </Tooltip>
      </div>

      <p>两种触发方式</p>
      <div className='demo-spin-1'>
        <Tooltip title='Tooltip!!!!!!!!!' trigger='click'>
          <Button>click</Button>
        </Tooltip>

        <Tooltip title='Tooltip!!!!!!!!!'>
          <Button>hover</Button>
        </Tooltip>
      </div>

      <p>外部visible 受控</p>
      <div>
        <Tooltip visible={visible} title='Tooltip!!!!!!!!!'>
          <Button onClick={() => setVisible(e => !e)}>受控</Button>
        </Tooltip>
      </div>

      <p>placement位置有 12 个方向</p>
      <div className='demo-tooltip-1'>
        {
          [topList, leftList, bottomList, rightList].map((e) => (
            <div key={e[1]} className={`demo-tooltip-${e[1]}`}>
              {
                e.map((ee) => (
                  <Tooltip key={ee} placement={(ee as any)} title='Tooltip!!!!!!!!!'>
                    <Button>{ee}</Button>
                  </Tooltip>
                ))
              }
            </div>
          ))
        }
      </div>

    </React.Fragment>
  )
}

export default Demo;
