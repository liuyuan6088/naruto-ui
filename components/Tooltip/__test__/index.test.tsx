import * as React from 'react'
import { mount } from 'enzyme'
import Tooltip from '../index'

describe('Tooltip', () => {
  it('render Tooltip', () => {
    const wrapper = mount(
      <Tooltip title="" transitionName="" mouseEnterDelay={0} mouseLeaveDelay={0}>
        <div id="hello">Hello world!</div>
      </Tooltip>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('Tooltip title is empty', () => {
    const onVisibleChange = jest.fn()

    const wrapper = mount(
      <Tooltip
        title=""
        transitionName=""
        mouseEnterDelay={0}
        mouseLeaveDelay={0}
        onVisibleChange={onVisibleChange}
      >
        <div id="hello">Hello world!</div>
      </Tooltip>
    )

    const div = wrapper.find('#hello').at(0)
    div.simulate('mouseenter')
    expect(onVisibleChange).not.toHaveBeenCalled()
  })
})
