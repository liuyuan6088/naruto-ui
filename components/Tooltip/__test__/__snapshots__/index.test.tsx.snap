// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip render Tooltip 1`] = `ReactWrapper {}`;
